ssl/t1_lib.c

```diff
@@ -2507,6 +2507,8 @@ static int has_usable_cert(SSL *s, const SIGALG_LOOKUP *sig, int idx)
                 || !X509_get_signature_info(s->cert->pkeys[idx].x509, &mdnid,
                                             &pknid, NULL, NULL))
                 continue;
+	    if (!EVP_PKEY_supports_digest_nid(s->cert->pkeys[idx].privatekey, mdnid))
+		    continue;
```

Comment:
I believe that given that the called functions can fail, `EVP_PKEY_supports_digest_nid()` cannot be a boolean function returning either `0` or `1`, but should allow also negative return values to signal errors from its callees.

This would also mean that this binary usage pattern cannot be applied.

Reply:
If `EVP_PKEY_supports_digest_nid()` is to go in the public API of `libcrypto` it needs to do the "right thing" and should not misuse other functions of the public API.
 
`libssl` is not the only user of `libcrypto`: _if_ it is ok for code in `libssl` to step over certain errors, it should do it by explicitly ignoring the error condition after checking the return value. Not doing so also exposes the library to potential pollution of the error stack.

Reply:
> No, checking `EVP_PKEY_supports_digest_nid()` for `==0` is all I care about. That means I cannot use the hash in question. If it returns 1, or if something else goes wrong, then I'll try the hash anyway. Because I certainly don't want to end up trying _no_ hashes, if something goes wrong here.

Anyway, in cases where this is acceptable --- and I believe it's not, given the motivation included in the previous comment --- when the return value of a function is not either 0 or 1, but can assume multiple values, we prefer to style the `if` by properly checking for a specific return value:
```C
if (EVP_PKEY_supports_digest_nid(s->cert->pkeys[idx].privatekey, mdnid) == 0) {
    /* do something */
}
```

Reply:
> Likewise, checking `EVP_PKEY_get_default_digest_nid()` for `==2` is all I care about, elsewhere. If there is a known mandatory hash, I honour it. If there is not a known mandatory hash (for whatever reason), then I do not. There is no "maybe".

IMO, this would apply if `EVP_PKEY_get_default_digest_nid()` would never fail, but given that the API clearly states that it might, it's an API misuse not to check the error condition.

You can still opt to not act upon the error and try anyway, but the code should clearly show this and motivate it, possibly checking if the error stack has been polluted.